common/src/main/scala/uk/ac/wellcome/s3/S3ObjectStore.scala

```diff
@@ -15,39 +15,55 @@ import scala.concurrent.Future
 import scala.io.Source
 import scala.util.hashing.MurmurHash3
 
-class SourcedObjectStore @Inject()(
+trait KeyPrefixGenerator[-T] {
+  def generate(obj: T): String
+}
+
+class SourcedKeyPrefixGenerator @Inject() extends KeyPrefixGenerator[Sourced] {
+  override def generate(obj: Sourced): String = {
+    val s3Shard = obj.sourceId.reverse.slice(0, 2)
+
+    s"${obj.sourceName}/${s3Shard}/${obj.sourceId}"
+  }
+}
+
+class S3ObjectStore[T] @Inject()(
   s3Client: AmazonS3,
-  @Flag("aws.s3.bucketName") bucketName: String)
-    extends Logging {
-  def put[T <: Sourced](sourcedObject: T)(
-    implicit encoder: Encoder[T]): Future[String] = {
+  @Flag("aws.s3.bucketName") bucketName: String,
+  keyPrefixGenerator: KeyPrefixGenerator[T]
+) extends Logging {
+  def put(sourcedObject: T)(implicit encoder: Encoder[T]): Future[String] = {
 
+    val keyPrefix = keyPrefixGenerator.generate(sourcedObject)
+    S3ObjectStore.put[T](s3Client, bucketName)(keyPrefix)(sourcedObject)
+  }
+
+  def get(key: String)(implicit decoder: Decoder[T]): Future[T] = {
+    S3ObjectStore.get[T](s3Client, bucketName)(key)
+  }
+}
+
+object S3ObjectStore extends Logging {
```

Comment:
Having both a class and an object of the same name feels like it's asking for trouble.

Maybe the class could be `KeyPrefixedS3ObjectStore`?

Reply:
This is  a "companion object" which is not unusual in Scala:  http://tutorials.jenkov.com/scala/singleton-and-companion-objects.html